src/main/java/edu/ie3/datamodel/io/source/sql/SqlTimeSeriesSource.java

```diff
@@ -199,6 +199,15 @@ public List<ZonedDateTime> getTimeKeysAfter(ZonedDateTime time) {
         .toList();
   }
 
+  @Override
+  public Optional<ZonedDateTime> getLastTimeKeyBefore(ZonedDateTime time) {
+    return dataSource
+        .executeQuery(
+            queryForValueBefore, ps -> ps.setTimestamp(1, Timestamp.from(time.toInstant())))
+        .map(valueFactory::extractTime)
+        .findFirst();
```

Comment:
is the stream ordered reverse or why does .findFirst() results in the last timeKey before?

Reply:
I improved the code, to address this point.